rust/main/agents/relayer/src/msg/metadata/base.rs

```diff
@@ -187,3 +192,189 @@ impl AppContextClassifier {
+            chains: Some(HashSet::from(["foochain".to_owned()])),
+            cache_policy: IsmCachePolicy::IsmSpecific,
+        };
+
+        assert_eq!(config.matches_chain("foochain"), true);
+        assert_eq!(config.matches_chain("barchain"), false);
+
+        assert_eq!(config.matches_module_type(ModuleType::Aggregation), true);
+        assert_eq!(config.matches_module_type(ModuleType::Routing), false);
+    }
+
+    #[test]
+    fn test_ism_cache_config_deserialize() {
+        // Module type 2 is the numeric version of ModuleType::Aggregation
+        let json = r#"
+        {
+            "moduletypes": [2],
+            "chains": ["foochain"],
+            "cachepolicy": "ismSpecific"
+        }
+        "#;
+
+        let config: IsmCacheConfig = serde_json::from_str(json).unwrap();
+
+        assert_eq!(
+            config.module_types,
+            HashSet::from([ModuleType::Aggregation])
+        );
+        assert_eq!(config.chains, Some(HashSet::from(["foochain".to_owned()])));
+        assert_eq!(config.cache_policy, IsmCachePolicy::IsmSpecific);
+    }
+
+    #[tokio::test]
+    async fn test_ism_cache_policy_classifier() {
+        let default_ism = H256::zero();
+
+        let mock_mailbox = MockMailboxContract::new_with_default_ism(default_ism);
+        let mailbox: Arc<dyn Mailbox> = Arc::new(mock_mailbox);
+
+        let default_ism_getter = DefaultIsmCache::new(mailbox);
+        let default_ism_cache_config = IsmCacheConfig {
+            module_types: HashSet::from([ModuleType::Aggregation]),
+            chains: Some(HashSet::from(["foochain".to_owned()])),
+            cache_policy: IsmCachePolicy::IsmSpecific,
+        };
+
+        let classifier =
+            IsmCachePolicyClassifier::new(default_ism_getter, default_ism_cache_config);
+
+        // We meet the criteria for the cache policy
+        let domain = HyperlaneDomain::new_test_domain("foochain");
+        let cache_policy = classifier
+            .get_cache_policy(default_ism, &domain, ModuleType::Aggregation)
+            .await;
+        assert_eq!(cache_policy, IsmCachePolicy::IsmSpecific);
+
+        // Different ISM module type, should not match
+        let cache_policy = classifier
+            .get_cache_policy(default_ism, &domain, ModuleType::Routing)
+            .await;
+        assert_eq!(cache_policy, IsmCachePolicy::MessageSpecific);
+
+        // ISM not default ISM, should not match
+        let cache_policy = classifier
+            .get_cache_policy(H256::repeat_byte(0xfe), &domain, ModuleType::Routing)
+            .await;
+        assert_eq!(cache_policy, IsmCachePolicy::MessageSpecific);
+
+        // Different domain, should not match
+        let domain = HyperlaneDomain::new_test_domain("barchain");
+        let cache_policy = classifier
+            .get_cache_policy(default_ism, &domain, ModuleType::Routing)
+            .await;
+        assert_eq!(cache_policy, IsmCachePolicy::MessageSpecific);
```

Comment:
wrote this before editing one of my comments above:

so once this PR is merged we will always be using cache? My initial understanding was that only `chains` in the `IsmCacheConfig` get to use cache, but seeing the result for `barchain` here shows otherwise. Would be nice to turn off caching if needed

Reply:
yeah good point, I'll add a config to globally allow caching

Reply:
added `allow_contract_call_caching` to make caching in general optional